module/zfs/zfs_vnops.c

```diff
@@ -862,11 +863,12 @@ zfs_write(struct inode *ip, uio_t *uio, int ioflag, cred_t *cr)
 		 * user 0 is not an ephemeral uid.
 		 */
 		mutex_enter(&zp->z_acl_lock);
+		retain_policy = ((zp->z_mode & S_ISUID) != 0 &&
+							zfs_uid_read(ip) == 0);
```

Comment:
Adding a local variable here doesn't really seem to buy us anything.  I'd suggest dropping it and updating the `zp->z_uid` check as you've done with the others.